app/src/main/java/org/lineageos/tv/launcher/adapter/TvAdapter.kt

```diff
@@ -28,38 +36,38 @@ abstract class TvAdapter<T : Card>(protected val context: Context) :
                 }
             }
         }
+
+        fun bind(item: L) {
+            card.setCardInfo(item)
+        }
     }
 
-    open fun handleClick(card: T) {
+    open fun handleClick(card: C) {
         val context = card.context
         context.startActivity(card.launchIntent)
         Toast.makeText(context, card.label, Toast.LENGTH_SHORT).show()
     }
 
-    open fun handleLongClick(card: T): Boolean {
-        return true
-    }
+    open fun handleLongClick(card: C) = false
 
     open fun handleKey(
-        card: T, keyCode: Int, event: KeyEvent?, bindingAdapterPosition: Int
+        card: C,
+        keyCode: Int,
+        event: KeyEvent?,
+        bindingAdapterPosition: Int,
     ) = false
 
-    open fun getAppsList() = AppManager.getInstalledApps(context)
-
-    override fun getItemCount() = launchablesList.size
+    companion object {
+        private fun <T : Launchable> getDiffCallback() = object : DiffUtil.ItemCallback<T>() {
+            override fun areItemsTheSame(
+                oldItem: T,
+                newItem: T,
+            ) = oldItem.packageName == newItem.packageName
 
-    fun removeItem(packageName: String) {
-        val index = launchablesList.indexOfFirst { it.packageName == packageName }
-        if (index != -1) {
-            launchablesList.removeAt(index)
-            notifyItemRemoved(index)
+            override fun areContentsTheSame(
+                oldItem: T,
+                newItem: T,
+            ) = oldItem.label == newItem.label && oldItem.icon == newItem.icon
```

Comment:
Drawables can't be compared.
Can use `oldItem.icon.toBitmap().sameAs(newItem.icon.toBitmap())` instead.

Reply:
https://gist.github.com/XinyueZ/3cca89416a1e443f914ed37f80ed59f2
Probably better, as it recycles the bitmap